pkg/cmd/cli/install/install.go

```diff
@@ -237,3 +257,41 @@ func (o *InstallOptions) Validate(c *cobra.Command, args []string, f client.Fact
 
 	return nil
 }
+
+// parseResourceRequests takes a set of CPU and memory requests and limit string
+// values and returns a ResourceRequirements struct to be used in a Container.
+// An error is returned if we cannot parse the request/limit.
+func parseResourceRequests(cpuRequest, memRequest, cpuLimit, memLimit string) (corev1.ResourceRequirements, error) {
+	var resources corev1.ResourceRequirements
+
+	parsedCPURequest, err := resource.ParseQuantity(cpuRequest)
```

Comment:
FYI it's possible to have an invalid CPU/memory request parsed successfully here that is later not accepted by the Kubernetes API. For example, a CPU value of "128Mi" would parse here (since it's a valid quantity for memory), but the API server will return the following error:

```
Error installing Velero. Use `kubectl logs deploy/velero -n velero` to check the deploy logs: Error creating resource Deployment/velero: Deployment.apps "velero" is invalid: spec.template.spec.containers[0].resources.requests: Invalid value: "5Mi": must be less than or equal to cpu limit
```

I think this is okay as the API server error is pretty clear, and there doesn't seem to be a way to parse CPU/memory values independently.

Reply:
Looks like we can parse them separately, but there's no helpers to do the requests <= limits check.

While I agree the message is fairly clear, I think it results in an incomplete install state - e.g. some things are created on the API server and some aren't. It's possible that this is already the case with some other errors during installation and not worth addressing here, though.

Reply:
@nrb hmm couldn't find how to parse them separately, can you point me to where you found that?

Reply:
Maybe I'm misunderstanding what you mean by separately, but this code is using `ParseQuantity` for each of the CPU/memory requests/limits (so 4 total calls for 4 total values).

Reply:
Sorry, what I meant is that I cannot parse just CPU quantities or just Memory quantities to avoid the issue where I could set CPU to "128Mi" which would be invalid.

Reply:
Ahh, I see. Got it.

Reply:
OK I see what's going on here.  Technically, "128Mi" isn't invalid for CPU, since there aren't actually units attached to these quantities - it just means a whole lot more CPUs than any machine actually has.

I think we should:
- *definitely* validate that the request <= the limit for both CPU and mem -- this will help avoid fat-finger errors
- *consider* adding reasonable upper bounds (e.g. CPU <= 8 (?), mem <= 64Gi (?)), again to catch blatant fat-finger errors (don't take my numbers as correct, just illustrative)

@nrb whatever we do here is probably relevant for #1677 as well, if we decide to allow user-configurable vals there

Reply:
Probably worth a function that we share to do the request <= limit comparison.

Reply:
Added a check for request <= limit, and added some tests around it.

> Probably worth a function that we share to do the request <= limit comparison.

 Currently just added it in the same parseResourceRequests function, do you think it makes sense to split it up?

Reply:
> consider adding reasonable upper bounds (e.g. CPU <= 8 (?), mem <= 64Gi (?)), again to catch blatant fat-finger errors (don't take my numbers as correct, just illustrative)

I guess this doesn't hurt, but I'm not sure how useful it would actually be. It's pretty clear when providing these flags what quantity (CPU or memory) is being set, and it might be a little arbitrary to define an upper bound. It might be better to check if the unit provided is acceptable for the quantity type (e.g. a simple check could just be to ensure the CPU quantity either has no unit, or the unit == "m")

Reply:
>Could #1677 make use of that whole function?

Yeah, it could.

Reply:
I think it depends what we're trying to protect against. To be clear -- "128Mi" is a valid request for CPU, as is "100m" for memory. The suffixes do not imply "cores" or "bytes", they're just shorthand for expressing unit-less numbers. However, a pod with a "128Mi" CPU request won't be scheduled unless you have a node with a hundred million+ CPUs :) The original error you posted wasn't complaining about using Mi for CPU, only about the fact that the request was bigger than the limit.

As long as we ensure request <= limit, then the pod will be accepted by the API server, though it may still not be scheduled if the requests are too high. This would be the case if the CPU request was, for example, "100", or "100Mi" (assuming no node with >=100 cores).

So if we just want to ensure the API server accepts the object, then I think we're good as-is. If we want to do some logical validation on the magnitude of the values, then we should decide what the boundaries of reasonable are there.

Reply:
`kubectl apply` this pod, then describe it:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: resource-request
spec:
  containers:
  - image: gcr.io/heptio-images/fsfreeze-pause:latest
    imagePullPolicy: Always
    name: fsfreeze
    resources:
      requests:
        cpu: 1Mi
  restartPolicy: Always
  serviceAccount: default
  serviceAccountName: default
```

Reply:
BTW, I think we're now fine as-is with the req<=lim check in place.

Reply:
> "128Mi" is a valid request for CPU, as is "100m" for memory. The suffixes do not imply "cores" or "bytes", they're just shorthand for expressing unit-less numbers.

Ah I see.

> So if we just want to ensure the API server accepts the object, then I think we're good as-is. If we want to do some logical validation on the magnitude of the values, then we should decide what the boundaries of reasonable are there.

Ultimately I think we want to result in a valid install, and ideally one that is not only accepted by the API server but is also configured correctly. That said, we can't guarantee that Velero will run with _any_ defined requests/limits (even the default) as we'd need to know what resources are available in the cluster, so I think there is a still an element of user intervention here to debug things if something went wrong. This is slightly different to a case where a user accidentally uses the wrong unit, but I think the end result is the same (pod can't be scheduled, user needs to debug at the Kubernetes level).

Reply:
Yep - and I do agree that any boundary we put in place will be arbitrary and may be problematic for a certain user.  Given that a `kubectl describe` will be pretty clear about there being insufficient resources to schedule the pod if the requests are too high, I think this is an acceptable UX.